parlai/tasks/igc/agents.py

```diff
@@ -0,0 +1,336 @@
+        """
+        return 2 * len(self.data)
+
+    def num_examples(self) -> int:
+        """
+        Number of examples.
+
+        There are three turns of dialogue in the IGC task -
+        Context, Question, Response.
+
+        Thus, return 3 * number of data examples.
+        """
+        return 3 * len(self.data)
+
+    def get(self, episode_idx: int, entry_idx: int = 0) -> dict:
+        """
+        Override to handle corrupt images and multi-reference labels.
+        """
+        entry_idx *= 2
+
+        if episode_idx >= len(self.data):
+            data = self.data[episode_idx % len(self.data)]
+            entry_idx += 1
+        else:
+            data = self.data[episode_idx]
+        image_id = data[self.image_id_key]
+        if data[self.image_id_key] not in self.valid_image_ids:
+            data[self.image_id_key] = self.blank_image_id
+        image_features = self.get_image_features(data)
+
+        conversation = [data['context'], data['question'], data['response']]
+        labels = [conversation[entry_idx]]
+        if self.multi_ref and entry_idx != 0:
+            key = 'questions' if entry_idx == 1 else 'responses'
+            labels = data[f'multiref_{key}'].split('***')
+        text = '' if entry_idx == 0 else conversation[entry_idx - 1]
+        episode_done = entry_idx >= len(conversation) - 2
+
+        action = {
+            'text': text,
+            'image_id': image_id,
+            'episode_done': episode_done,
+            'image': image_features,
+            'labels': labels,
+        }
+
+        return action
+
+    def load_data(self, data_path: str, opt: Dict[str, Any]) -> List[Dict[str, Any]]:
+        """
+        Override to load CSV files.
+        """
+
+        dt = opt['datatype'].split(':')[0]
+        dt_str = 'test' if dt == 'test' else 'val'
+        dp = os.path.join(self.get_data_path(opt), f'IGC_crowd_{dt_str}.csv')
+        if not os.path.exists(dp):
+            raise RuntimeError(
+                'Please download the IGC Dataset from '
+                'https://www.microsoft.com/en-us/download/details.aspx?id=55324. '
+                'Then, make sure to put the two .csv files in ParlAI/data/igc/'
+            )
+        if (
+            not os.path.exists(self.get_image_path(opt))
+            or len(os.listdir(self.get_image_path(opt))) <= 1
+        ):
+            self._download_images(opt)
+
+        self.data = []
+        with open(dp, newline='\n') as csv_file:
+            reader = csv.reader(csv_file, delimiter=',')
+            fields = []
+            for i, row in enumerate(reader):
+                if i == 0:
+                    fields = row
+                else:
+                    ep = dict(zip(fields, row))
+                    ep['image_id'] = f'{ep["id"]}'
+                    self.data.append(ep)
+
+        if dt == 'train':
+            # Take first 90% of valid set as train
+            self.data = self.data[: int(len(self.data) * 0.9)]
```

Comment:
is the data pre-shuffled?

Reply:
to my knowledge no, however there doesn't appear to be any sense of order in the datafile. this is an arbitrary choice anyway, as there is no train set really...